### ISSUE TITLE
Add variable requesting to NotebookControllers

### ISSUE THREAD

**Issue by rchiodo** [microsoft/vscode#133114]
Jupyter python notebooks support this view here:

![image](https://user-images.githubusercontent.com/19672699/133313926-3775d13a-21ff-408a-8270-0a39bfdd5464.png)

In order to facilitate sharing this view (or creating a view in VS code core that does the same thing), this issue is proposing some additions to notebook controllers

Edited (from comments below):
- ```requestVariables(notebook: NotebookDocument, controller: NotebookController, args: VariablesArguments): Thenable<Variable[]>```
- ```requestVariableColumns(notebook: NotebookDocument, controller: NotebookController): Thenable<string []>```

With the idea being:
- ```requestVariables``` is called to fill out a page of variables. Types come from the [variables](https://microsoft.github.io/debug-adapter-protocol/specification#Requests_Variables) request of the DAP.
- ```requestVariableColumns``` is called to retrieve column headers. [Variable (from DAP)](https://microsoft.github.io/debug-adapter-protocol/specification#Types_Variable) could be added to in order to support more than just value, type, and name.

We should probably also think about:
- Returning other mimetypes for variable display other than just plain text.

**Comment by DonJayamanne** [comment]
How similar or different would these views be compared to the variable viewer when debugging?
I feel we could end up duplicate UIs when a user is debugging (variables in variable viewer & also in debug viewer), agree this is already a UX issue in jupyter today, but we'd now be extending this to other languages & other notebooks as well.

Personally I'd like to see an API similar to that of the DAP (which supports drilling into nested objects - I believe this is what Julia does), which isn't tied to a specific UX such as a table view (the current API seems hardcoded to a specific UX)

**Comment by DonJayamanne** [comment]
Pinging Julia folk @davidanthoff /cc

**Comment by rchiodo** [comment]
Yeah I agree a object based version would make sense too. Hence why I suggested we look at returning object graphs too.

That would be like (following what DAP has)

```
interface Variable {
  /**
   * The variable's name.
   */
  name: string;

  /**
   * The variable's value. This can be a multi-line text, e.g. for a function
   * the body of a function.
   */
  value: string;

  /**
   * The type of the variable's value. Typically shown in the UI when hovering
   * over the value.
   * This attribute should only be returned by a debug adapter if the client has
   * passed the value true for the 'supportsVariableType' capability of the
   * 'initialize' request.
   */
  type?: string;

  /**
   * Properties of a variable that can be used to determine how to render the
   * variable in the UI.
   */
  presentationHint?: VariablePresentationHint;

  /**
   * Optional evaluatable name of this variable which can be passed to the
   * 'EvaluateRequest' to fetch the variable's value.
   */
  evaluateName?: string;

  /**
   * If variablesReference is > 0, the variable is structured and its children
   * can be retrieved by passing variablesReference to the VariablesRequest.
   */
  variablesReference: number;

  /**
   * The number of named child variables.
   * The client can use this optional information to present the children in a
   * paged UI and fetch them in chunks.
   */
  namedVariables?: number;

  /**
   * The number of indexed child variables.
   * The client can use this optional information to present the children in a
   * paged UI and fetch them in chunks.
   */
  indexedVariables?: number;

  /**
   * Optional memory reference for the variable if the variable represents
   * executable code, such as a function pointer.
   * This attribute is only required if the client has passed the value true for
   * the 'supportsMemoryReferences' capability of the 'initialize' request.
   */
  memoryReference?: string;
 
 /**
  * Additional columns to be shown. Key is from requestColumns request
  */
  additionalColumns: { [key: string]: any };
}
```

I think that would change the requests like so:
```requestVariableReferences(notebook: NotebookDocument, controller: NotebookController, start: number, count: number): Thenable<number[]>```
```requestVariableColumns(notebook: NotebookDocument, controller: NotebookController): Thenable<string []>```
```requestVariable(notebook: NotebookDocument, controller: NotebookController, variableReference: number): Thenable<string []>```

**Comment by IanMatthewHuff** [comment]
Can it be more exact to the DAP? Like use VariablesRequest with VariablesArguments directly instead of having something like start and count as parameters? I would think something on the notebook controller would want to be both well defined and flexible, so using the DAP Requests straight up sounds pretty good to me.

Plus it's a step on the way to us just hosting or utilizing the VS Code variables control directly.

**Comment by rchiodo** [comment]
> Can it be more exact to the DAP? Like use VariablesRequest with VariablesArguments directly instead of having something like start and count as parameters? I would think something on the notebook controller would want to be both well defined and flexible, so using the DAP Requests straight up sounds pretty good to me.
> 
> Plus it's a step on the way to us just hosting or utilizing the VS Code variables control directly.

Yeah sure. So you mean this:

```requestVariables(notebook: NotebookDocument, controller: NotebookController, args: VariablesArguments): Thenable<Variable[]>```
```requestVariableColumns(notebook: NotebookDocument, controller: NotebookController): Thenable<string []>```
[]>```

**Comment by IanMatthewHuff** [comment]
Yeah, that's a bit what I was thinking. Just making it as similar to the DAP as possible felt better to me.

**Comment by davidanthoff** [comment]
We had a very productive call with @claudiaregio last week, so I think she has a good handle on what we would need from the Julia side of things. @LadyNaggaga also participated (I hope I got the right github handle!) and it looked like the .Net and Julia requirements were pretty well aligned, I'd say?

One crucial requirement for us is that a variable explorer is not exclusive tied to a notebook controller, we also use the variable explorer for our REPL, and for us to swap our own version out for something shared would require that it supports that scenario as well. The other major thing is support for drill down, we currently use a protocol that is very much modeled after the DAP structure, so anything that mimics that would probably work for us.

There are a couple of other more minor things, but I think we talked through all of these in our call.

CC @pfitzseb
